File: src/main/java/cub/hackerrank/tutorials/a30DaysOfCode/day28/day28.py
#!/usr/bin/env python3

import math
import os
import random
import re
import sys


class Row:
    def __init__(self, firstName, emailID):
        self.firstName = firstName
        self.emailID = emailID


if __name__ == '__main__':
    db = list()
    N = int(input())
    for N_itr in range(N):
        firstNameEmailID = input().split()
        firstName = firstNameEmailID[0]
        emailID = firstNameEmailID[1]
        db.append(Row(firstName, emailID))

    gmail_regexp = re.compile('^.*@gmail.com$')
    for firstName in sorted(map(lambda row: row.firstName, filter(lambda row: gmail_regexp.match(row.emailID), db))):
        print(firstName)

File: src/main/java/cub/hackerrank/tutorials/a30DaysOfCode/day28/day28.pl
#!/usr/bin/env perl

use strict;
use warnings;

my $N = <>;
$N =~ s/\s+$//;

my $db = [];

for (my $N_itr = 0; $N_itr < $N; $N_itr++) {
  my $firstNameEmailID = <>;
  $firstNameEmailID =~ s/\s+$//;
  my @firstNameEmailID = split /\s+/, $firstNameEmailID;

  my $firstName = $firstNameEmailID[0];
  chomp($firstName);

  my $emailID = $firstNameEmailID[1];
  chomp($emailID);

  push(@$db, [ $firstName, $emailID ]);
}

print join("\n" => sort {$a cmp $b}
  map {$_->[0]}
    grep {$_->[1] =~ m/^.*\@gmail.com$/}
      @$db
)
